### PR TITLE
add list of open PRs which updated a given out-of-sync file.

### PR DESCRIPTION
**Pull request by joneugster**
Add another column to the "out-of-sync" page that lists all open Mathlib4 PRs with label mathlib3-pair that touch the out-of-sync file.

Moreover move remaining Github API calls (for labels) to the mathlib4-script `make_port_status.py`

- [x] depends on: leanprover-community/mathlib4#2520 

[Open in Gitpod](https://gitpod.io/#https://github.com/joneugster/mathlib-port-status/)

### CODE REVIEW

**Comment by eric-wieser** [comment]
I recommend you create a new branch for this rather than PRing from `main`; that way you won't end up discarding this again.

File: port_status_yaml.py [hunk 1]
```diff
@@ -18,10 +18,17 @@ class Comment:
     class Source:
         repo: str
         commit: str
+    @dataclass
+    class Label:
+        name: str
+        color: str
+        text_color: str = '000000'
```
Comment (eric-wieser) [inline port_status_yaml.py]: I don't think `text_color` belongs here; we don't need it in the `yaml`, and this file should act as a specification for what's actually in the yaml file.

File: port_status_yaml.py [hunk 2]
```diff
@@ -18,10 +18,17 @@ class Comment:
...
     ported: bool
     source: Optional[Source]
     mathlib4_pr: Optional[int]
     mathlib4_file: Optional[str]
+    labels: Optional[list[Label]]
+    sync_prs: Optional[list[int]]
```
Comment (eric-wieser) [inline port_status_yaml.py]: These aren't optional (in the python meaning of "can be `None`"), but if absent the semantics are an empty list
```suggestion
    labels: list[Label] = field(default_factory=list)
    sync_prs: list[int] = field(default_factory=list)
```

File: templates/out-of-sync.j2 [hunk 3]
```diff
@@ -106,9 +107,16 @@ $(document).ready(function () {
                 <span class="text-success" title="added">+{{added}}</span>&nbsp;<span class="text-danger" title="removed">-{{removed}}</span>
             </td>
             <td>{{ d.dependent_depth }}</td>
+            <td>
```
Comment (eric-wieser) [inline templates/out-of-sync.j2]:
```suggestion
            <td data-order="{{len(d.status.sync_prs)}}">
```

Comment (joneugster) [reply templates/out-of-sync.j2]: or `min` to see the oldest PR?

File: port_status_yaml.py (hunk 1 above)
Comment (joneugster) [reply port_status_yaml.py]: It's just I don't know how to call a python function in the `.j2` file, so I need to assign `label.text_color = ...` somewhere.

Comment (eric-wieser) [reply port_status_yaml.py]: You can add the function to `template_env.filters` (see elsewhere in the file), and then apply it as `label.color | text_color`.

File: templates/out-of-sync.j2 (hunk 3 above)
Comment (eric-wieser) [reply templates/out-of-sync.j2]: That will error if there are no PRs. I don't think age (measured by PR number) is interesting, whereas being able to spot duplicates is.

File: templates/out-of-sync.j2 [hunk 4]
```diff
@@ -106,9 +107,16 @@ $(document).ready(function () {
                 <span class="text-success" title="added">+{{added}}</span>&nbsp;<span class="text-danger" title="removed">-{{removed}}</span>
             </td>
             <td>{{ d.dependent_depth }}</td>
+            <td data-order="{{d.status.mathlib4_sync_prs | min}}">
```
Comment (eric-wieser) [inline templates/out-of-sync.j2]:
```suggestion
            <td data-order="{{d.status.mathlib4_sync_prs | length}}">
```
as discussed in another review thread

File: templates/out-of-sync.j2 [hunk 5]
```diff
@@ -106,9 +107,16 @@ $(document).ready(function () {
...
+                <ul>
+                    {%- for pr in d.status.mathlib4_sync_prs %}
+                        <li><a href="https://github.com/leanprover-community/mathlib4/pull/{{pr}}">#{{pr}}</a></li>
+                    {%- endfor %}
+                </ul>
```
Comment (eric-wieser) [inline templates/out-of-sync.j2]: Can you add a similar list near the top of the "Changes since the initial port" section of the `file.j2`? You could probably include the labels there.

File: port_status_yaml.py (hunk 1 above)
Comment (joneugster) [reply port_status_yaml.py]: nvm, done.

File: templates/out-of-sync.j2 (hunk 5 above)
Comment (joneugster) [reply templates/out-of-sync.j2]: Done

File: make_html.py [hunk 6]
```diff
@@ -364,7 +343,10 @@ def make_index(env, html_root):
         index_f.write(
             env.get_template('index.j2').render(
                 all=data.values(),
-                ported=ported, unported=unported, in_progress=in_progress))
+                ported=ported,
+                unported=unported,
+                in_progress=in_progress,
+                text_color_of_color=text_color_of_color))
```
Comment (eric-wieser) [inline make_html.py]: You should put this in `template_env.filters` insteead, see line 263.

File: templates/file.j2 [hunk 7]
```diff
@@ -52,6 +52,24 @@
...
+{% if data.status.mathlib4_sync_prs %}
+    <div class="alert alert-warning">
+        This file has open sync PRs:
+        <ul>
+            {%- for pr in data.status.mathlib4_sync_prs %}
+                <li><a href="https://github.com/leanprover-community/mathlib4/pull/{{pr.num}}">#{{pr.num}}</a>
+                    {%- for label in pr.labels %}
+                        <a class="badge rounded-pill"
+                            href='https://github.com/leanprover-community/mathlib4/pulls?q=is%3Apr+is%3Aopen+label%3A"{{label.name}}"'
+                            style="background-color:#{{label.color}};color:{{text_color_of_color(label.color)}}">
```
Comment (eric-wieser) [inline templates/file.j2]:
```suggestion
                            style="background-color:#{{label.color}};color:{{ label.color | text_color_of_color}}">
```
after making it a filter

**Comment by eric-wieser** [comment]
Can you enable actions in your fork so that I can see a preview? I would imagine there's an obvious button at https://github.com/joneugster/mathlib-port-status/actions that you can enable them with.

**Comment by joneugster** [comment]
> Can you enable actions in your fork so that I can see a preview? I would imagine there's an obvious button at https://github.com/joneugster/mathlib-port-status/actions that you can enable them with.

I pressed that obvious button, but I've never done anything with actions... it says "0 workflows". Let me see...

File: make_html.py [hunk 8]
```diff
@@ -163,9 +138,8 @@ class Mathlib3FileData:
     mathlib3_import: List[str]
     status: port_status_yaml.PortStatusEntry
     lines: Optional[int]
-    labels: Optional[List[dict[str, str]]]
-    dependents: Optional[List['Mathlib3FileData']] = None
-    dependencies: Optional[List['Mathlib3FileData']] = None
+    dependents: List['Mathlib3FileData'] = field(default_factory=list)
+    dependencies: List['Mathlib3FileData'] = field(default_factory=list)
```
Comment (eric-wieser) [inline make_html.py]: This is inaccurate; these can be `None` if the file doesn't have any data at all, as opposed to importing nothing.

**Comment by eric-wieser** [comment]
> > Can you enable actions in your fork so that I can see a preview? I would imagine there's an obvious button at https://github.com/joneugster/mathlib-port-status/actions that you can enable them with.
> 
> I pressed that obvious button, but I've never done anything with actions... it says "0 workflows". Let me see...

https://github.com/joneugster/mathlib-port-status/actions/workflows/pages.yml now should have a "run" button in the top right.

File: make_html.py (hunk 8 above)
Comment (joneugster) [reply make_html.py]: Hmm, I got an error in `file.j2` at l.259 (`{{ progress.bars([data] + data.dependencies) }}`) that `data.dependencies` is `None` but expected to be `list`. I'll fix that another way then.

Comment (eric-wieser) [reply make_html.py]: This usually means that you need to `git pull` in your `mathlib` directory.

But ideally we would add a `{% if data.dependencies is none %}some warning message{%else%} the progress bar {% endif %}`

**Comment by eric-wieser** [comment]
> https://github.com/joneugster/mathlib-port-status/actions/workflows/pages.yml now should have a "run" button in the top right.

You'll now need to turn on pages in the repository settings.

File: make_html.py (hunk 8 above)
Comment (joneugster) [reply make_html.py]: > This usually means that you need to `git pull` in your `mathlib` directory.

Ah

**Comment by eric-wieser** [comment]
> You'll now need to turn on pages in the repository settings.

... and set it to use "github actions" instead of the other option.

**Comment by joneugster** [comment]
I think I have that now, Im sorry

File: templates/out-of-sync.j2 [hunk 9]
```diff
@@ -106,9 +107,17 @@ $(document).ready(function () {
                 <span class="text-success" title="added">+{{added}}</span>&nbsp;<span class="text-danger" title="removed">-{{removed}}</span>
             </td>
             <td>{{ d.dependent_depth }}</td>
+            <td data-order="{{d.status.mathlib4_sync_prs | length}}">
+                <ul>
+                    {%- for pr in d.status.mathlib4_sync_prs %}
+                        <li><a href="https://github.com/leanprover-community/mathlib4/pull/{{pr.num}}"
+                            title="{%- for label in pr.labels %}{{label.name}}{{ ", " if not loop.last }}{%- endfor %}">#{{pr.num}}</a></li>
+                    {%- endfor %}
+                </ul>
```
Comment (eric-wieser) [inline templates/out-of-sync.j2]: Given we expect to rarely have more than one, I think it would be better to just use separate lines than waste space with bullets
```suggestion
                {% set separator = joiner('<br />') %}
                {%- for pr in d.status.mathlib4_sync_prs %}{{ separator() }}
                    <a href="https://github.com/leanprover-community/mathlib4/pull/{{pr.num}}"
                        title="{{ pr.labels | map(attribute="name") | join(", ") }}">#{{pr.num}}</a>
                {%- endfor %}
```

File: make_html.py [hunk 10]
```diff
@@ -364,7 +344,10 @@ def make_index(env, html_root):
         index_f.write(
             env.get_template('index.j2').render(
                 all=data.values(),
-                ported=ported, unported=unported, in_progress=in_progress))
+                ported=ported,
+                unported=unported,
+                in_progress=in_progress,
+                text_color_of_color=text_color_of_color))
```
Comment (eric-wieser) [inline make_html.py:350]:
```suggestion
                in_progress=in_progress))
```

Comment (eric-wieser) [inline make_html.py]:
```suggestion
                ported=ported, unported=unported, in_progress=in_progress))
```

File: templates/file.j2 [hunk 11]
```diff
@@ -46,12 +46,30 @@
...
+     style="background-color:#{{label.color}};color:{{label.color | text_color_of_color}}">
     {{label.name}}
   </a>
   {%- endfor %}
 </div>
 {% endif %}
+{% if data.status.mathlib4_sync_prs %}
+    <div class="alert alert-warning">
+        This file has open sync PRs:
+        <ul>
```
Comment (eric-wieser) [inline templates/file.j2]: I think this section would go well immediately under the "changes in mathlib4" heading. I plan to augment it with the commit information in a later date.

File: templates/file.j2 [hunk 12]
```diff
@@ -238,43 +256,47 @@ cards
...
-    })
-    </script>
-    <p>The unported dependencies are</p>
-    <ul>
-        {% for d in unported_deps %}
-            <li><a href="{{ site_url }}/file/{{d.mathlib3_import|join('/')}}"><code>{{d.mathlib3_import|join('.')}}</code></a></li>
-        {% endfor %}
-    </ul>
+{% if data.dependencies is none %}
+    Could not display informations. You might need to update your copy of `mathlib`.
```
Comment (eric-wieser) [inline templates/file.j2]:
```suggestion
    <div class="alert alert-danger">Dependency information not available. This file might be missing from the version of mathlib this website was built against.</div>
```

File: make_html.py [hunk 13]
```diff
@@ -446,6 +426,7 @@ def make_out_of_sync(env, html_root, mathlib_dir):
                     mathlib4_import=mathlib4_import,
                     data=get_data().get(f_import),
                     graph=graph,
+                    text_color_of_color=text_color_of_color,
```
Comment (eric-wieser) [inline make_html.py]:
```suggestion
```

File: templates/file.j2 [hunk 14]
```diff
@@ -238,43 +254,47 @@ cards
...
-    })
-    </script>
-    <p>The unported dependencies are</p>
-    <ul>
-        {% for d in unported_deps %}
-            <li><a href="{{ site_url }}/file/{{d.mathlib3_import|join('/')}}"><code>{{d.mathlib3_import|join('.')}}</code></a></li>
-        {% endfor %}
-    </ul>
+{% if data.dependencies is none %}
+    <div class="alert alert-danger">Dependency information not available. This file might be missing from the version of mathlib this website was built from.</div>
```
Comment (eric-wieser) [inline templates/file.j2]: I split this out into a separate commit that's now in master; if you merge and solve the conflicts, the diff in this section should go away.

File: templates/file.j2 [hunk 15]
```diff
@@ -173,6 +173,22 @@ if (range != null) {
...
+                {% if data.status.mathlib4_sync_prs %}
+                    <div class="alert alert-warning">
+                        This file has an open sync PR:
+                        {%- for pr in data.status.mathlib4_sync_prs %}
+                            <br />
+                            <a href="https://github.com/leanprover-community/mathlib4/pull/{{pr.num}}">#{{pr.num}}</a>
+                            {%- for label in pr.labels %}
+                                <a class="badge rounded-pill"
+                                    href='https://github.com/leanprover-community/mathlib4/pulls?q=is%3Apr+is%3Aopen+label%3A"{{label.name}}"'
+                                    style="background-color:#{{label.color}};color:{{text_color_of_color(label.color)}}">
```
Comment (eric-wieser) [inline templates/file.j2]:
```suggestion
                                    style="background-color:#{{label.color}};color:{{label.color | text_color_of_color}}">
```

File: templates/file.j2 [hunk 16]
```diff
@@ -54,7 +54,7 @@
   {%- for label in data.labels %}
```
Comment (eric-wieser) [inline templates/file.j2]: This isn't working any more; I assume it should be
```suggestion
  {%- for label in data.status.labels %}
```

**Review by eric-wieser — APPROVED** [review]
Thanks, this is great!

**Comment by joneugster** [comment]
Thanks for the help! :)

File: templates/out-of-sync.j2 [hunk 17]
```diff
@@ -106,9 +107,16 @@ $(document).ready(function () {
                 <span class="text-success" title="added">+{{added}}</span>&nbsp;<span class="text-danger" title="removed">-{{removed}}</span>
             </td>
             <td>{{ d.dependent_depth }}</td>
+            <td data-order="{{d.status.mathlib4_sync_prs | length}}">
+                {% set separator = joiner('<br />') %}
```
Comment (eric-wieser) [inline templates/out-of-sync.j2:111]: I'm a bit suspicious that jinja is going to HTML-escape this, but it's hard to tell right now as we don't have any duplicate forward-port PRs!

Comment (eric-wieser) [reply templates/out-of-sync.j2:111]: Fixed in https://github.com/leanprover-community/mathlib-port-status/commit/1fc623179ec6754015cb5fc641834fc736bbe991
